File: src/models/quest.py
"""
Quest model - Represents a task or goal in the QuestBoard system
"""

from datetime import datetime
from enum import Enum
from typing import Optional, List
from dataclasses import dataclass, field


class QuestType(Enum):
    """Types of quests available"""
    BRAINSTORM = "Brainstorm"
    BUILD = "Build"
    PROMOTE = "Promote"
    DEBUG = "Debug"
    LEARN = "Learn"
    REVIEW = "Review"
    DOCUMENTATION = "Documentation"


class QuestDifficulty(Enum):
    """Quest difficulty levels"""
    EASY = "Easy"
    MEDIUM = "Medium"
    HARD = "Hard"
    EPIC = "Epic"


class QuestStatus(Enum):
    """Quest status lifecycle"""
    CREATED = "Created"
    ASSIGNED = "Assigned"
    IN_PROGRESS = "In Progress"
    COMPLETED = "Completed"
    ABANDONED = "Abandoned"


@dataclass
class Quest:
    """
    Represents a quest (task/goal) in the QuestBoard system.
    
    Attributes:
        title: Short, descriptive quest name
        description: Detailed quest information
        difficulty: Quest difficulty level
        quest_type: Category of quest
        xp_reward: Experience points awarded upon completion
        status: Current quest status
        assigned_to: Username of assigned team member
        created_at: Timestamp of quest creation
        completed_at: Timestamp of quest completion
        tags: Optional tags for categorization
    """
    title: str
    description: str
    difficulty: QuestDifficulty
    quest_type: QuestType
    xp_reward: int
    status: QuestStatus = field(default=QuestStatus.CREATED)
    assigned_to: Optional[str] = None
    created_at: datetime = field(default_factory=datetime.now)
    completed_at: Optional[datetime] = None
    tags: List[str] = field(default_factory=list)
    quest_id: str = field(default_factory=lambda: f"quest_{datetime.now().timestamp()}")
    
    def __post_init__(self):
        """Validate quest data after initialization"""
        if not self.title or len(self.title.strip()) == 0:
            raise ValueError("Quest title cannot be empty")
        
        if self.xp_reward < 0:
            raise ValueError("XP reward must be non-negative")
        
        # Convert string enums to proper enum types if needed
        if isinstance(self.difficulty, str):
            self.difficulty = QuestDifficulty[self.difficulty.upper()]
        
        if isinstance(self.quest_type, str):
            self.quest_type = QuestType[self.quest_type.upper().replace(" ", "_")]
        
        if isinstance(self.status, str):
            self.status = QuestStatus[self.status.upper().replace(" ", "_")]
    
    def assign_to(self, username: str) -> None:
        """
        Assign the quest to a team member.
        
        Args:
            username: Username of the team member
        """
        if not username or len(username.strip()) == 0:
            raise ValueError("Username cannot be empty")
        
        self.assigned_to = username
        if self.status == QuestStatus.CREATED:
            self.status = QuestStatus.ASSIGNED
    
    def start(self) -> None:
        """Mark the quest as in progress"""
        if self.status not in [QuestStatus.ASSIGNED, QuestStatus.CREATED]:
            raise ValueError(f"Cannot start quest with status: {self.status.value}")
        
        self.status = QuestStatus.IN_PROGRESS
    
    def complete(self) -> int:
        """
        Mark the quest as completed and return XP reward.
        
        Returns:
            XP reward amount
        """
        if self.status == QuestStatus.COMPLETED:
            raise ValueError("Quest is already completed")
        
        self.status = QuestStatus.COMPLETED
        self.completed_at = datetime.now()
        return self.xp_reward
    
    def abandon(self) -> None:
        """Mark the quest as abandoned"""
        if self.status == QuestStatus.COMPLETED:
            raise ValueError("Cannot abandon a completed quest")
        
        self.status = QuestStatus.ABANDONED
    
    def get_icon(self) -> str:
        """Get the emoji icon for this quest type"""
        icons = {
            QuestType.BRAINSTORM: "🧠",
            QuestType.BUILD: "🔧",
            QuestType.PROMOTE: "📣",
            QuestType.DEBUG: "🐛",
            QuestType.LEARN: "📚",
            QuestType.REVIEW: "👀",
            QuestType.DOCUMENTATION: "📝",
        }
        return icons.get(self.quest_type, "⭐")
    
    def to_dict(self) -> dict:
        """Convert quest to dictionary representation"""
        return {
            'quest_id': self.quest_id,
            'title': self.title,
            'description': self.description,
            'difficulty': self.difficulty.value,
            'quest_type': self.quest_type.value,
            'xp_reward': self.xp_reward,
            'status': self.status.value,
            'assigned_to': self.assigned_to,
            'created_at': self.created_at.isoformat(),
            'completed_at': self.completed_at.isoformat() if self.completed_at else None,
            'tags': self.tags,
            'icon': self.get_icon(),
        }
    
    def __str__(self) -> str:
        """String representation of the quest"""
        return f"{self.get_icon()} [{self.difficulty.value}] {self.title} ({self.xp_reward} XP)"

File: questboard_app/lib/models/quest.dart
import 'package:hive/hive.dart';

part 'quest.g.dart';

@HiveType(typeId: 2)
enum QuestType {
  @HiveField(0)
  brainstorm,
  @HiveField(1)
  build,
  @HiveField(2)
  promote,
  @HiveField(3)
  research,
  @HiveField(4)
  test,
  @HiveField(5)
  review,
  @HiveField(6)
  learn,
}

@HiveType(typeId: 3)
enum QuestDifficulty {
  @HiveField(0)
  easy,
  @HiveField(1)
  medium,
  @HiveField(2)
  hard,
  @HiveField(3)
  epic,
}

@HiveType(typeId: 4)
enum QuestStatus {
  @HiveField(0)
  todo,
  @HiveField(1)
  inProgress,
  @HiveField(2)
  completed,
  @HiveField(3)
  blocked,
  @HiveField(4)
  cancelled,
}

@HiveType(typeId: 5)
class Quest extends HiveObject {
  @HiveField(0)
  final String id;

  @HiveField(1)
  final String title;

  @HiveField(2)
  final String description;

  @HiveField(3)
  final QuestType type;

  @HiveField(4)
  final QuestDifficulty difficulty;

  @HiveField(5)
  final int xpReward;

  @HiveField(6)
  final QuestStatus status;

  @HiveField(7)
  final String assigneeId;

  @HiveField(8)
  final String teamId;

  @HiveField(9)
  final DateTime createdAt;

  @HiveField(10)
  final DateTime? dueDate;

  @HiveField(11)
  final DateTime? completedAt;

  @HiveField(12)
  final List<String> tags;

  @HiveField(13)
  final List<QuestStep> steps;

  @HiveField(14)
  final String? githubBranch;

  @HiveField(15)
  final List<String> attachmentUrls;

  @HiveField(16)
  final int estimatedHours;

  @HiveField(17)
  final List<String> dependsOn;

  Quest({
    required this.id,
    required this.title,
    required this.description,
    required this.type,
    required this.difficulty,
    required this.xpReward,
    this.status = QuestStatus.todo,
    required this.assigneeId,
    required this.teamId,
    required this.createdAt,
    this.dueDate,
    this.completedAt,
    this.tags = const [],
    this.steps = const [],
    this.githubBranch,
    this.attachmentUrls = const [],
    this.estimatedHours = 1,
    this.dependsOn = const [],
  });

  Quest copyWith({
    String? id,
    String? title,
    String? description,
    QuestType? type,
    QuestDifficulty? difficulty,
    int? xpReward,
    QuestStatus? status,
    String? assigneeId,
    String? teamId,
    DateTime? createdAt,
    DateTime? dueDate,
    DateTime? completedAt,
    List<String>? tags,
    List<QuestStep>? steps,
    String? githubBranch,
    List<String>? attachmentUrls,
    int? estimatedHours,
    List<String>? dependsOn,
  }) {
    return Quest(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      type: type ?? this.type,
      difficulty: difficulty ?? this.difficulty,
      xpReward: xpReward ?? this.xpReward,
      status: status ?? this.status,
      assigneeId: assigneeId ?? this.assigneeId,
      teamId: teamId ?? this.teamId,
      createdAt: createdAt ?? this.createdAt,
      dueDate: dueDate ?? this.dueDate,
      completedAt: completedAt ?? this.completedAt,
      tags: tags ?? this.tags,
      steps: steps ?? this.steps,
      githubBranch: githubBranch ?? this.githubBranch,
      attachmentUrls: attachmentUrls ?? this.attachmentUrls,
      estimatedHours: estimatedHours ?? this.estimatedHours,
      dependsOn: dependsOn ?? this.dependsOn,
    );
  }

  double get progress {
    if (steps.isEmpty) return status == QuestStatus.completed ? 1.0 : 0.0;
    final completedSteps = steps.where((step) => step.isCompleted).length;
    return completedSteps / steps.length;
  }

  static int getXPForDifficulty(QuestDifficulty difficulty) {
    switch (difficulty) {
      case QuestDifficulty.easy:
        return 15;
      case QuestDifficulty.medium:
        return 35;
      case QuestDifficulty.hard:
        return 75;
      case QuestDifficulty.epic:
        return 150;
    }
  }

  static String getColorForType(QuestType type) {
    switch (type) {
      case QuestType.brainstorm:
        return '#FF6B35';
      case QuestType.build:
        return '#4ECDC4';
      case QuestType.promote:
        return '#45B7D1';
      case QuestType.research:
        return '#96CEB4';
      case QuestType.test:
        return '#FFEAA7';
      case QuestType.review:
        return '#DDA0DD';
      case QuestType.learn:
        return '#FFB6C1';
    }
  }
}

@HiveType(typeId: 6)
class QuestStep extends HiveObject {
  @HiveField(0)
  final String id;

  @HiveField(1)
  final String title;

  @HiveField(2)
  final bool isCompleted;

  @HiveField(3)
  final DateTime? completedAt;

  @HiveField(4)
  final String? notes;

  QuestStep({
    required this.id,
    required this.title,
    this.isCompleted = false,
    this.completedAt,
    this.notes,
  });

  QuestStep copyWith({
    String? id,
    String? title,
    bool? isCompleted,
    DateTime? completedAt,
    String? notes,
  }) {
    return QuestStep(
      id: id ?? this.id,
      title: title ?? this.title,
      isCompleted: isCompleted ?? this.isCompleted,
      completedAt: completedAt ?? this.completedAt,
      notes: notes ?? this.notes,
    );
  }
}
